boskos/common/common.go

```diff
@@ -19,37 +19,173 @@ package common
+func NewResource(name, rtype, state, owner string, t time.Time) Resource {
+	return Resource{
+		Name:       name,
+		Type:       rtype,
+		State:      state,
+		Owner:      owner,
+		LastUpdate: t,
+		UserData:   map[string]string{},
+	}
+}
+
+// NewResourcesFromConfig parse the a ResourceEntry into a list of resources
+func NewResourcesFromConfig(e ResourceEntry) []Resource {
+	var resources []Resource
+	for _, name := range e.Names {
+		resources = append(resources, NewResource(name, e.Type, e.State, "", time.Time{}))
+	}
+	return resources
+}
+
+// UserDataNotFound will be returned if requested resource does not exist.
+type UserDataNotFound struct {
+	ID string
+}
+
+func (ud *UserDataNotFound) Error() string {
+	return fmt.Sprintf("user data ID %s does not exist", ud.ID)
 }
 
+// ResourceByUpdateTime helps sorting resources by update time
+type ResourceByUpdateTime []Resource
+
+func (ut ResourceByUpdateTime) Len() int           { return len(ut) }
+func (ut ResourceByUpdateTime) Swap(i, j int)      { ut[i], ut[j] = ut[j], ut[i] }
+func (ut ResourceByUpdateTime) Less(i, j int) bool { return ut[i].LastUpdate.Before(ut[j].LastUpdate) }
+
+// ResourceByName helps sorting resources by name
+type ResourceByName []Resource
+
+func (ut ResourceByName) Len() int           { return len(ut) }
+func (ut ResourceByName) Swap(i, j int)      { ut[i], ut[j] = ut[j], ut[i] }
+func (ut ResourceByName) Less(i, j int) bool { return ut[i].GetName() < ut[j].GetName() }
+
+// ResTypes is used to parse flags for a list of types
 type ResTypes []string
 
 func (r *ResTypes) String() string {
 	return fmt.Sprint(*r)
 }
 
-func (rtypes *ResTypes) Set(value string) error {
-	if len(*rtypes) > 0 {
+// Set parses the flag value into a ResTypes
+func (r *ResTypes) Set(value string) error {
+	if len(*r) > 0 {
 		return errors.New("resTypes flag already set")
 	}
 	for _, rtype := range strings.Split(value, ",") {
-		*rtypes = append(*rtypes, rtype)
+		*r = append(*r, rtype)
 	}
 	return nil
 }
+
+// GetName implements the Item interface used for storage
+func (res Resource) GetName() string { return res.Name }
+
+// Extract unmarshalls a string a given struct if it exists
+func (ud UserData) Extract(id string, out interface{}) error {
+	content, ok := ud[id]
+	if !ok {
+		return &UserDataNotFound{id}
+	}
+	return yaml.Unmarshal([]byte(content), out)
+}
+
+// User Data are used to store custom information mainly by Mason and Masonable implementation.
+// Mason used a LeasedResource keys to store information about other resources that used to
+// create the given resource.
+
+// Set marshalls a struct to a string into the UserData
+func (ud UserData) Set(id string, in interface{}) error {
+	b, err := yaml.Marshal(in)
+	if err != nil {
+		return err
+	}
+	ud[id] = string(b)
+	return nil
+}
+
+// Update updates existing UserData with new UserData.
+// If a key as an empty string, the key will be deleted
+func (ud UserData) Update(new UserData) {
+	if new != nil {
```

Comment:
I think this would be bad, and user might be deleting user data they don't know about. In order delete userdata you need to send an update of a key with an empty string. I think it is better if it s explicit.